### PR TITLE
refactor(mysql): port to sqlglot

### PR DESCRIPTION
**Pull request by cpcloud**
Port the MySQL backend to sqlglot. Depends on #7877.

### CODE REVIEW

File: ibis/backends/mssql/tests/test_client.py [hunk 1]
```diff
@@ -65,10 +60,6 @@
 @pytest.mark.parametrize(
     ("server_type", "expected_type"),
     DB_TYPES
-    + [
-        param("GEOMETRY", dt.geometry, marks=[skipif_no_geospatial_deps]),
-        param("GEOGRAPHY", dt.geography, marks=[skipif_no_geospatial_deps]),
-    ]
```
Comment (cpcloud) [inline ibis/backends/mssql/tests/test_client.py:71]: I'll bring this back in the MS SQL port

File: ibis/backends/mysql/converter.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+                second=comps.seconds,
+                microsecond=comps.microseconds,
+            )
+
+        return s.map(convert, na_action="ignore")
+
+    @classmethod
+    def convert_Timestamp(cls, s, dtype, pandas_type):
+        if s.dtype == "object":
+            s = s.replace("0000-00-00 00:00:00", None)
```
Comment (cpcloud) [inline ibis/backends/mysql/converter.py:25]: This existed before but as a `result_processor` method on a custom sqlalchemy type. This is a good example of something that goes through many fewer layers of software by removing sqlalchemy from the mix here.

File: ibis/backends/base/sqlglot/__init__.py [hunk 3]
```diff
@@ -296,7 +296,7 @@ def to_pyarrow_batches(
         schema = expr.as_table().schema()
         array_type = schema.as_struct().to_pyarrow()
         arrays = (
-            pa.array(map(tuple, batch), type=array_type)
+            pa.array(map(tuple, batch), type=array_type, safe=False)
```
Comment (cpcloud) [inline ibis/backends/base/sqlglot/__init__.py:299]: Revert this.

File: ibis/backends/tests/test_numeric.py [hunk 4]
```diff
@@ -254,7 +255,7 @@ def test_numeric_literal(con, backend, expr, expected_types):
                 "postgres": decimal.Decimal("1.1"),
                 "pandas": decimal.Decimal("1.1"),
                 "pyspark": decimal.Decimal("1.1"),
-                "mysql": 1.1,
+                "mysql": decimal.Decimal("1"),
```
Comment (cpcloud) [inline ibis/backends/tests/test_numeric.py:258]: This isn't `"1.1"` due to the behavior of MySQL's default decimal parameters. See the sqlglot type mapper for MySQL for details.

File: ibis/expr/rewrites.py [hunk 5]
```diff
@@ -92,6 +92,48 @@ def rewrite_dropna(_):
     return ops.Filter(_.parent, tuple(preds))
 
 
+@replace(p.WindowFunction(p.First(x, y)))
```
Comment (cpcloud) [inline ibis/expr/rewrites.py:95]: Delete these and use the ones in the sqlglot base ruleset.

Comment (cpcloud) [reply ibis/expr/rewrites.py]: Done.

File: ibis/backends/base/sqlglot/__init__.py (hunk 3 above)
Comment (cpcloud) [reply ibis/backends/base/sqlglot/__init__.py]: Done.

**Comment by cpcloud** [comment]
I'll remove the geospatial function mapping since we're not testing that for mysql yet.

File: ibis/backends/tests/test_numeric.py (hunk 4 above)
Comment (kszucs) [reply ibis/backends/tests/test_numeric.py:258]: The ibis datatype of 
```py
ibis.literal(decimal.Decimal("1.1"), type=dt.decimal).type()
```
is
```
Decimal(precision=None, scale=None, nullable=True)
```

so the literal is casted to `DECIMAL` and MySQL decides about scale and precision. 

We should have another test case where we test a decimal literal with explicit scale and precision, and we expect uniform behaviour across all the backends.

Comment (kszucs) [reply ibis/backends/tests/test_numeric.py:258]: Created a follow-up https://github.com/ibis-project/ibis/issues/7930

File: ibis/formats/pandas.py [hunk 6]
```diff
@@ -150,12 +150,15 @@ def convert_table(cls, df, schema):
...
         pandas_type = PandasType.from_ibis(dtype)
 
-        if obj.dtype == pandas_type and dtype.is_primitive():
-            return obj
-
         method_name = f"convert_{dtype.__class__.__name__}"
         convert_method = getattr(cls, method_name, cls.convert_default)
 
+        if convert_method is cls.convert_default:
+            # only take the fast path if it isn't overridden in a subclass
```
Comment (kszucs) [inline ibis/formats/pandas.py:157]: Then the fast path itself should be implemented in `convert_default()` instead.

Comment (cpcloud) [reply ibis/formats/pandas.py]: Done.

**Review by kszucs — APPROVED** [review]

**Comment by cpcloud** [comment]
Snowflake tests look good:

```
…/ibis on  tes-mysql is 📦 v7.2.0 via 🐍 v3.10.13 via ❄️  impure (ibis-3.10.13-env)
❯ pytest -m snowflake --snapshot-update -n 8 --dist loadgroup -q
bringing up nodes...
...x..x.........................x..x.x.x....xxx......x.x....x...x....x.x.....xxxxx...xx..xx...x.x...x.x.xx............x..x....x.x..xx.xx...x.x.....x.x.......xx..x...xx....x.x....x......x.xx. [ 13%]
.x......x..x.......x....x......x....xx....x.x...........x.....xxxxx........x........x.....x..............................................x..x......xxx....xx.x.....xx..........x.............. [ 26%]
...x......x....x........................x.x.......x..........x.x.....x......................s..x.........xx.x.x...x.xxx....xxxxxxx...........x.x..x........x......x.....x.x.x.........x....... [ 39%]
...........x.........x................x.x.......x........x....x........x..................xxx........................x........x..........x.x.......x....x.....xx...................xx......x.. [ 52%]
...xx.xxx.x..xx....x...x...x......x.....................xx...x....x..x.x.....x...............x.....................xx...............x.......x.............x........x.....x..x...x..........x.. [ 65%]
........xx........x..............x.....x..........xxx......x.x...........x.............x..........x...........x..................................x.............................x.............x [ 78%]
..x.....x...xxxx.x...x.......x....x..............x.....x.............................................................................s........................................................ [ 91%]
................................................................................................................................s..                                                            [100%]
1260 passed, 3 skipped, 198 xfailed in 138.45s (0:02:18)
```

**Pull request merged**